### PR TITLE
flang: add support for llvm-6

### LINKED ISSUE #8072: Flang update to use llvm-6
currently only `llvm@:5.9999` works!

But build instructions have changed:
https://github.com/flang-compiler/flang/commit/cbd29f8d7b3903c12d44ff8f6489373f9aaad852#diff-04c6e90faac2675aa89e2176d2eec7d8

/cc @MarquitoForrest

### PR DESCRIPTION
**Pull request by junghans**
Fix #8072 

/cc @MarquitoForrest @sscalpone

### CODE REVIEW

**Comment by MarquitoForrest** [comment]
I will try out your PR on Friday! Thanks :=)

**Comment by junghans** [comment]
@aprokop

**Comment by junghans** [comment]
Needs #8272

**Comment by aprokop** [comment]
I'm very confused. What's the difference between `flang` and `flang-driver`? Does `flang-driver` no longer requires clang as it's forked from it?

**Comment by aprokop** [comment]
And this still needs #8276 as current `develop` is broken.

**Comment by junghans** [comment]
`flang-driver` is the new approach and it needs a patched llvm (https://github.com/llvm-mirror/llvm/pull/56)

**Comment by aprokop** [comment]
The first and only issue in `flang-driver` says:
```
Do not fork or build from this repository. After discussions with the Flang community, we thought we would try to create a fork of the LLVM Clang repository to house the changes necessary for Flang, analogous to what we are doing with the mirrored llvm repository.

Creating the fork named flang-driver is the first step in that process.
```

**Comment by junghans** [comment]
Oh man, looks like a mess, so we need a `llvm` and a `clang` patch, plus `flang` itself.

**Comment by junghans** [comment]
We have the `llvm` patch, but we are basically missing: https://github.com/llvm-mirror/clang/compare/release_60...flang-compiler:release_60.

**Comment by junghans** [comment]
@aprokop that should hopefully work!

**Comment by junghans** [comment]
@aprokop work for me, please test!

@adamjstewart: please review!

File: var/spack/repos/builtin/packages/flang/package.py [hunk 1]
```diff
@@ -72,17 +79,30 @@ def post_install(self):
...
-            out.write(
-                '{0} -I{1} -L{2} {3}{4} -B{5} "$@"\n'.format(
-                    os.path.join(self.spec['llvm'].prefix.bin, 'flang'),
-                    self.prefix.include, self.prefix.lib,
-                    self.compiler.fc_rpath_arg, self.prefix.lib,
-                    self.spec.prefix.bin))
+            if '@develop' in self.spec:
+                out.write(
+                    '{0} -I{1} -L{2} -L{3} {4}{5} {6}{7} -B{8} "$@"\n'.format(
+                        os.path.join(self.spec['llvm'].prefix.bin, 'flang'),
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/flang/package.py]: You can actually use `self.spec.prefix.bin.flang` here, no need for `os.path.join`. Same elsewhere in the `package.py`

File: var/spack/repos/builtin/packages/pgmath/package.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+##############################################################################
+
+from spack import *
+import os
+
+
+class Pgmath(CMakePackage):
+    """Flang's math library"""
+    homepage = "https://github.com/flang-compiler/flang"
+
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/pgmath/package.py]: You might encounter problems uninstalling this package without a fake URL. This is a known bug in Spack.

File: var/spack/repos/builtin/packages/flang/package.py (hunk 1 above)
Comment (junghans) [reply var/spack/repos/builtin/packages/flang/package.py]: Fixed

File: var/spack/repos/builtin/packages/pgmath/package.py (hunk 2 above)
Comment (junghans) [reply var/spack/repos/builtin/packages/pgmath/package.py]: Done!

**Comment by junghans** [comment]
@adamjstewart, yeah I might to a PR to flang to support and all-in-one build, but for now it is easier to make it a separate package. I guess the idea is that `pgmath` could be used to replace the math library of other fortran compilers.

**Comment by aprokop** [comment]
sorry, I have not had time to test it out, and probably won't this week. What happened to `flang-driver` I remember seeing in this PR?

**Comment by junghans** [comment]
@aprokop that is https://github.com/llvm-mirror/clang/pull/40!

**Comment by junghans** [comment]
@aprokop ping

**Pull request merged**